modules/auxiliary/scanner/http/gitlab_graphql_user_enum.rb

```diff
@@ -0,0 +1,180 @@
+
+    user_json = res.get_json_document
+    fail_with(Failure::UnexpectedReply, "The target didn't return a JSON body") if user_json.nil?
+
+    nodes = user_json.dig('data', 'users', 'nodes')
+    fail_with(Failure::UnexpectedReply, 'Could not find nodes in the JSON body') if nodes.nil?
+
+    user_json
+  end
+
+  ##
+  # Parses the JSON data returned by the server. Adds the usernames to
+  # the users array and adds them, indirectly, to create_credential_login.
+  # This function also determines if we need to request more data from
+  # the server.
+  #
+  # @param user_json [Hash] The JSON data provided by the server
+  # @param users [Array] An array to store new usernames in
+  # @return [String] An empty string or the "endCursor" to use with do_request
+  ##
+  def parse_json(user_json, users)
+    nodes = user_json.dig('data', 'users', 'nodes')
+    return '' if nodes.nil?
+
+    nodes.each do |node|
+      username = node['username']
+      store_username(username, node)
+      users.push(username)
+    end
+
+    query_paging_info = ''
+    more_data = user_json.dig('data', 'users', 'pageInfo', 'hasNextPage')
+    if !more_data.nil? && more_data == true
+      query_paging_info = user_json['data']['users']['pageInfo']['endCursor']
+    end
+
+    query_paging_info
+  end
+
+  def store_userlist(users, service)
+    loot = store_loot('gitlab.users', 'text/plain', rhost, users, nil, 'GitLab Users', service)
+    print_good("Userlist stored at #{loot}")
+  end
+
+  def store_username(username, json)
+    service = ssl ? 'https' : 'http'
+    service_data = {
+      address: rhost,
+      port: rport,
+      service_name: service,
+      protocol: 'tcp',
+      workspace_id: myworkspace_id,
+      proof: json
+    }
+
+    credential_data = {
+      origin_type: :service,
+      module_fullname: fullname,
+      username: username
+    }
+
+    credential_data.merge!(service_data)
+
+    # Create the Metasploit::Credential::Core object
+    credential_core = create_credential(credential_data)
+
+    # Assemble the options hash for creating the Metasploit::Credential::Login object
+    login_data = {
+      core: credential_core,
+      status: Metasploit::Model::Login::Status::UNTRIED
+    }
+
+    # Merge in the service data and create our Login
+    login_data.merge!(service_data)
+    create_credential_login(login_data)
+  end
+
+  ##
+  # Send an initial GraphQL request to the server and keep sending
+  # requests until the server has no more data to give us.
+  ##
+  def run_host(_ip)
+    user_json = do_request('')
+
+    service = report_service(
+      host: rhost,
+      port: rport,
+      name: (ssl ? 'https' : 'http'),
+      proto: 'tcp'
+    )
+
+    # parse the initial page
+    users = Array[]
+    query_paging_info = parse_json(user_json, users)
+
+    # handle any follow on pages
+    until query_paging_info.empty?
+      user_json = do_request(query_paging_info)
```

Comment:
Not a blocker: I wonder if it's worth adding progress output here

Reply:
Do we have some type of builtin logic for that @adfoster-r7? I'm happy to add it if so!

Reply:
I believe scanners normally log out progress - but it's not super reusable here:

https://github.com/rapid7/metasploit-framework/blob/02143c5142be135d7dc1080a38e631b6a41b3d40/lib/msf/core/auxiliary/scanner.rb#L308-L317

Maybe just a log message every $x requests saying that we're still doing things would be useful? Not a blocker for me - feel free to punt/resolve the thread!

```
request_count = 0
until query_paging_info.empty?
  # Wording open to improvement
  print_status("Enumerating users - request #{request_count}") if request_count % 10 == 0

  ...
end
```

Reply:
I added a print_status for every 5th pagination request. Because GitLab returns 100 users to us at a time, this is unlikely to appear except in big orgs, but doesn't hurt! The test case I have in documentation (142 users) doesn't hit this case because all are returned in two requests, but I artificially lowered the modulus to print every page request and it looked like so:

```
msf6 auxiliary(scanner/http/gitlab_graphql_user_enum) > run

[*] GraphQL API pagination request: 1
[+] Enumerated 142 GitLab users
[+] Userlist stored at /home/albinolobster/.msf4/loot/20220311071819_default_10.0.0.13_gitlab.users_852540.txt
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```